### PR TITLE
Add Ejection, ejections, and shiftEjection for destructing sums

### PR DESCRIPTION
**Pull request by adamse**

### CODE REVIEW

File: sop-core/src/Data/SOP/NS.hs [hunk 1]
```diff
@@ -156,6 +161,33 @@ instance All (NFData `Compose` f) xs => NFData (NS f xs) where
     rnf (Z x)  = rnf x
     rnf (S xs) = rnf xs
 
+-- | The type of ejections from an n-ary sum.
+--
+-- An ejection is the pattern matching function for one part of the n-ary sum.
+--
+-- @since 0.5.0.0 -- TODO: correct version?
```
Comment (adamse) [inline sop-core/src/Data/SOP/NS.hs]: Is this the version to put here?

**Comment by kosmikus** [comment]
Thanks!

It would certainly make sense to include this in the library.

Some remarks:

1. I'm not convinced of the name "ejection" for the concept, but I also do not have a better suggestion right now. Is "ejection" established terminology anywhere (if so, could you point me to it?), or did you invent the name? The closest relationship in Haskell land so far seems to be prisms, so I'm wondering if we should just call it `prisms` rather than `ejections`, because it might provide better intuition to what it does?

2. I think I'd prefer targetting `Maybe` specifically rather than an arbitrary `Alternative`. One can always go from `Maybe` to any `Alternative`, but using a concrete type is usually clearer.

3. Don't worry about the version numbers right now. I indeed think that the most likely next release is 0.5.0.0 because I want to merge the metadata change, but we can fix these once we know this.

4. I don't much like LambdaCase and would prefer if you could rewrite without it. There are a few other stylistical changes I would probably make, but I'm also in principle happy to apply them myself after merging.

**Comment by adamse** [comment]
1. Ejection due it being the opposite of an injection. I see didn't make this connection in the docs but this is why I choose the name.
2. Happy to make it just Maybe, this is the only type I use it at so far too.
3. Ok
4. Will remove LambdaCase

**Comment by kosmikus** [comment]
I've never been able to come up with a better name. I'll just merge this. Sorry for letting it linger for so long. Thanks!

**Pull request merged**